src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
```

Comment:
Instead of using multiple `if` conditions, you should use a `switch` statement to handle different `action.type` values as it is more suitable for this kind of limited, known set of values. Remember to include a `default` case in your `switch` statement for error handling.